File: php/test/OrganizationTeamEntityTest.php
<?php
declare(strict_types=1);

// OrganizationTeam entity test

require_once __DIR__ . '/../cloudsmith_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class OrganizationTeamEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = CloudsmithSDK::test(null, null);
        $ent = $testsdk->OrganizationTeam(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = organization_team_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "organization_team." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set CLOUDSMITH_TEST_ORGANIZATION_TEAM_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $organization_team_ref01_ent = $client->OrganizationTeam(null);
        $organization_team_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.organization_team"), "organization_team_ref01"));
        $organization_team_ref01_data["org_id"] = $setup["idmap"]["org01"];

        $organization_team_ref01_data_result = $organization_team_ref01_ent->create($organization_team_ref01_data, null);
        $organization_team_ref01_data = Helpers::to_map($organization_team_ref01_data_result);
        $this->assertNotNull($organization_team_ref01_data);

        // LIST
        $organization_team_ref01_match = [
            "org_id" => $setup["idmap"]["org01"],
        ];

        $organization_team_ref01_list_result = $organization_team_ref01_ent->list($organization_team_ref01_match, null);
        $this->assertIsArray($organization_team_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($organization_team_ref01_list_result),
            ["id" => $organization_team_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $organization_team_ref01_data_up0_up = [
            "org_id" => $setup["idmap"]["org_id"],
        ];

        $organization_team_ref01_markdef_up0_name = "description";
        $organization_team_ref01_markdef_up0_value = "Mark01-organization_team_ref01_" . $setup["now"];
        $organization_team_ref01_data_up0_up[$organization_team_ref01_markdef_up0_name] = $organization_team_ref01_markdef_up0_value;

        $organization_team_ref01_resdata_up0_result = $organization_team_ref01_ent->update($organization_team_ref01_data_up0_up, null);
        $organization_team_ref01_resdata_up0 = Helpers::to_map($organization_team_ref01_resdata_up0_result);
        $this->assertNotNull($organization_team_ref01_resdata_up0);
        $this->assertEquals($organization_team_ref01_resdata_up0[$organization_team_ref01_markdef_up0_name], $organization_team_ref01_markdef_up0_value);

        // LOAD
        $organization_team_ref01_match_dt0 = [];
        $organization_team_ref01_data_dt0_loaded = $organization_team_ref01_ent->load($organization_team_ref01_match_dt0, null);
        $this->assertNotNull($organization_team_ref01_data_dt0_loaded);

    }
}

function organization_team_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/organization_team/OrganizationTeamTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = CloudsmithSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["organization_team01", "organization_team02", "organization_team03", "org01", "org02", "org03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("CLOUDSMITH_TEST_ORGANIZATION_TEAM_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "CLOUDSMITH_TEST_ORGANIZATION_TEAM_ENTID" => $idmap,
        "CLOUDSMITH_TEST_LIVE" => "FALSE",
        "CLOUDSMITH_TEST_EXPLAIN" => "FALSE",
        "CLOUDSMITH_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["CLOUDSMITH_TEST_ORGANIZATION_TEAM_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }
    if (!isset($idmap_resolved["org_id"])) {
        $idmap_resolved["org_id"] = $idmap_resolved["org01"];
    }

    if ($env["CLOUDSMITH_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["CLOUDSMITH_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new CloudsmithSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["CLOUDSMITH_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["CLOUDSMITH_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: go/test/organization_team_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/cloudsmith-sdk/go"
	"github.com/voxgig-sdk/cloudsmith-sdk/go/core"

	vs "github.com/voxgig-sdk/cloudsmith-sdk/go/utility/struct"
)

func TestOrganizationTeamEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.OrganizationTeam(nil)
		if ent == nil {
			t.Fatal("expected non-nil OrganizationTeamEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"organization_team": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.OrganizationTeam(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.OrganizationTeam(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := organization_teamBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "organization_team." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set CLOUDSMITH_TEST_ORGANIZATION_TEAM_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		organizationTeamRef01Ent := client.OrganizationTeam(nil)
		organizationTeamRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "organization_team"}, setup.data), "organization_team_ref01"))
		organizationTeamRef01Data["org_id"] = setup.idmap["org01"]

		organizationTeamRef01DataResult, err := organizationTeamRef01Ent.Create(organizationTeamRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		organizationTeamRef01Data = core.ToMapAny(organizationTeamRef01DataResult)
		if organizationTeamRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LIST
		organizationTeamRef01Match := map[string]any{
			"org_id": setup.idmap["org01"],
		}

		organizationTeamRef01ListResult, err := organizationTeamRef01Ent.List(organizationTeamRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		organizationTeamRef01List, organizationTeamRef01ListOk := organizationTeamRef01ListResult.([]any)
		if !organizationTeamRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", organizationTeamRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(organizationTeamRef01List), map[string]any{"id": organizationTeamRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		organizationTeamRef01DataUp0Up := map[string]any{
			"org_id": setup.idmap["org_id"],
		}

		organizationTeamRef01MarkdefUp0Name := "description"
		organizationTeamRef01MarkdefUp0Value := fmt.Sprintf("Mark01-organization_team_ref01_%d", setup.now)
		organizationTeamRef01DataUp0Up[organizationTeamRef01MarkdefUp0Name] = organizationTeamRef01MarkdefUp0Value

		organizationTeamRef01ResdataUp0Result, err := organizationTeamRef01Ent.Update(organizationTeamRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		organizationTeamRef01ResdataUp0 := core.ToMapAny(organizationTeamRef01ResdataUp0Result)
		if organizationTeamRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if organizationTeamRef01ResdataUp0[organizationTeamRef01MarkdefUp0Name] != organizationTeamRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", organizationTeamRef01MarkdefUp0Name, organizationTeamRef01ResdataUp0[organizationTeamRef01MarkdefUp0Name])
		}

		// LOAD
		organizationTeamRef01MatchDt0 := map[string]any{}
		organizationTeamRef01DataDt0Loaded, err := organizationTeamRef01Ent.Load(organizationTeamRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if organizationTeamRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func organization_teamBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "organization_team", "OrganizationTeamTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read organization_team test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse organization_team test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"organization_team01", "organization_team02", "organization_team03", "org01", "org02", "org03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("CLOUDSMITH_TEST_ORGANIZATION_TEAM_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"CLOUDSMITH_TEST_ORGANIZATION_TEAM_ENTID": idmap,
		"CLOUDSMITH_TEST_LIVE":      "FALSE",
		"CLOUDSMITH_TEST_EXPLAIN":   "FALSE",
		"CLOUDSMITH_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["CLOUDSMITH_TEST_ORGANIZATION_TEAM_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add org_id alias for update test.
	if idmapResolved["org_id"] == nil {
		idmapResolved["org_id"] = idmapResolved["org01"]
	}

	if env["CLOUDSMITH_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["CLOUDSMITH_APIKEY"],
			},
			extra,
		})
		client = sdk.NewCloudsmithSDK(core.ToMapAny(mergedOpts))
	}

	live := env["CLOUDSMITH_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["CLOUDSMITH_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
